distributed/client.py

```diff
@@ -3530,22 +3530,36 @@ def restart(self, timeout=no_default, wait_for_workers=True):
         )
 
     async def _restart_workers(
-        self, workers: list[str], timeout: int | float | None = None
+        self,
+        workers: list[str],
+        timeout: int | float | None = None,
+        raise_for_timeout: bool = True,
     ) -> dict[str, str]:
-        results: dict[str, str] = await self.scheduler.broadcast(
-            msg={"op": "restart", "timeout": timeout}, workers=workers, nanny=True
+        info = self.scheduler_info()
+        name_to_addr = {meta["name"]: addr for addr, meta in info["workers"].items()}
+        worker_addrs = [name_to_addr.get(w, w) for w in workers]
+
+        restart_out: dict[str, str] = await self.scheduler.broadcast(
+            msg={"op": "restart", "timeout": timeout},
+            workers=worker_addrs,
+            nanny=True,
         )
-        timeout_workers = {
-            key: value for key, value in results.items() if value == "timed out"
-        }
-        if timeout_workers:
+
+        # Map keys back to original `workers` input names/addresses
+        results = {w: restart_out[w_addr] for w, w_addr in zip(workers, worker_addrs)}
+
+        timeout_workers = [w for w, status in results.items() if status == "timed out"]
+        if timeout_workers and raise_for_timeout:
             raise TimeoutError(
-                f"The following workers failed to restart with {timeout} seconds: {list(timeout_workers.keys())}"
+                f"The following workers failed to restart with {timeout} seconds: {timeout_workers}"
             )
         return results
 
     def restart_workers(
-        self, workers: list[str], timeout: int | float | None = None
+        self,
+        workers: list[str],
+        timeout: int | float | None = None,
+        raise_for_timeout: bool = True,
```

Comment:
maybe we should rather use `raise_on_error` since there are cases where the problem is not just a timeout.